test/metabase/test/util.clj

```diff
@@ -282,21 +282,24 @@
   "Temporarily set the value of the Setting named by keyword `setting-k` to `value` and execute `f`, then re-establish
   the original value. This works much the same way as `binding`.
 
-   Prefer the macro `with-temporary-setting-values` over using this function directly."
+  Prefer the macro `with-temporary-setting-values` over using this function directly."
   {:style/indent 2}
   [setting-k value f]
   ;; plugins have to be initialized because changing `report-timezone` will call driver methods
   (initialize/initialize-if-needed! :db :plugins)
-  (let [setting        (#'setting/resolve-setting setting-k)
-        original-value (when (or (#'setting/db-or-cache-value setting)
-                                 (#'setting/env-var-value setting))
-                         (setting/get setting-k))]
+  (let [setting                    (#'setting/resolve-setting setting-k)
+        env-var-value              (#'setting/env-var-value setting)
+        original-db-or-cache-value (#'setting/db-or-cache-value setting)]
     (try
       (setting/set! setting-k value)
       (testing (colorize/blue (format "\nSetting %s = %s\n" (keyword setting-k) (pr-str value)))
-        (f))
+        (if env-var-value
+          (testing (colorize/red (format "NOTE: %s is set to %s by an env var, which takes precedence\n"
+                                         (keyword setting-k) (pr-str env-var-value)))
+            (f))
+          (f)))
       (finally
-        (setting/set! setting-k original-value)))))
+        (setting/set! setting-k original-db-or-cache-value)))))
```

Comment:
This function originally had a bug, if a setting wasn't saved in the DB but was set via an env var:
- The original value was fetched with `setting/get`, which would return the env var value
- This value would then be written to the DB in the `finally` form

I guess this never was an issue with the old precedence rules, since usually the DB value was being returned by `setting/get` anyway, but with the new rules it could be fairly confusing. I think this function should only ever read and write the DB value, since `do-with-temp-env-var-value` is a separate helper.

I've instead added some logging if you try to use this function on a setting that is also set by an env var.

Reply:
I think the fact that `do-with-temporary-setting-values` won't override the Setting value in certain cases will indeed be a little confusing. How about we just have it check whether the Setting was set via an env var and call `do-with-temp-env-var-value` if it was? That way `do-with-temporary-setting-values` will always do the right thing, and you won't have to be aware of special rules to know when it will and won't work.

Then you'd only need to use `with-temp-env-var-value` for changing env vars that aren't related to Settings

Reply:
Either way, if you were to make this change, that little gotcha is definitely something worth calling out in the docstring for `with-temporary-setting-values`. Otherwise you'd either have to notice the `testing` context (which you might not see, if your test is (possibly incorrectly) passing), or go in and read the definition of the implementation function for the macro to learn about an important usage caveat

Reply:
Yeah, I guess that makes sense -- I'll update the PR to reflect that behavior.